### PR TITLE
feat: add ability to create string conditions

### PR DESCRIPTION
**Pull request by phui01**
#### Is this adding or improving a _feature_ or fixing a _bug_?
Adding a new feature.

#### What's the new behavior?
Exact string matching

#### How does this change work?
As shown in numerous videos posted to, and discussed over, Teams.

### Does this fix any issues or need any specific reviewers?

Fixes: #
Reviewers: @

### CODE REVIEW

File: packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx [hunk 1]
```diff
@@ -174,15 +174,13 @@ class Container extends React.Component<Props, ComponentState> {
                         ? NONE_RESOLVER_KEY
                         : nextProps.entity.resolverType
 
-                    const isMultiValueDisabled = this.isMultiValueDisabled(nextProps.entity, this.props.actions)
```
Comment (mattmazzola) [inline packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx]: Think you need more complex logic here other than simply removing this function.

Changing to be non multivalue would have still allowed conditions that are invalid to be exist which might could cause errors.  Do you remove all those conditions on actions?  Seems easier to block but I guess you can ask @swadheen

File: packages/ui/src/components/modals/ConditionModal.tsx [hunk 2]
```diff
@@ -50,9 +49,15 @@ const convertConditionTypesToDropdownOptions = (conditionTypes: object): Operato
...
 
-const operatorOptions = convertConditionTypesToDropdownOptions(CLM.ConditionType)
+const allOperatorOptions = convertConditionTypesToDropdownOptions(CLM.ConditionType)
 // We know EQUAL will be found since it was created from enum type in line above
-const equalOperatorOption = operatorOptions.find(o => o.data === CLM.ConditionType.EQUAL)!
+const equalOperatorOption = allOperatorOptions.find(o => o.data === CLM.ConditionType.EQUAL)!
+const stringEqualOperatorOption = allOperatorOptions.find(o => o.data === CLM.ConditionType.STRING_EQUAL)!
+const stringEqualOperatorOptions = allOperatorOptions.filter(o => o.data === CLM.ConditionType.STRING_EQUAL)!
+const arithmeticOperatorOptions = allOperatorOptions.filter(o => o.data !== CLM.ConditionType.STRING_EQUAL)!
+var operatorOptions = allOperatorOptions
```
Comment (mattmazzola) [inline packages/ui/src/components/modals/ConditionModal.tsx]: `var`... 

It seems weird for this variable to not be state managed, but perhaps thats ordering issue.  Would check that.

Also, the `operatorOptions` likely isn't needed.
Based on the UI the user already constraints the operator types to three sets:

Entity Class | Comparison Type | Operators | Value
--- | --- | --- | ---
enum | string | == | "string"
multivalue | number | [==, !=, >, <, etc] | 234
custom, has number resolver | number | [==, !=, >, <, etc] | 98
.   | string | == | "my value"
(all other classes) | string | == | "my string"

I this case there is never a need to "change" the available operators to the DropDown control, you simply select it and make it disabled if it's constrained to equality (such as equals)

I thought the introduction of "comparison type" eliminated the need to have second equality operator.
image

I would prefer that say "==" but not sure what you and @LarsLiden decided.

Also, when you select enum there seems to be bug with comparison type being undefined. Enums work based on "string" value equality so I thought it should say "string".  I'm not sure if this was intentional but think users might wonder. 

image

**Comment by mattmazzola** [comment]
Also, you should update the title of PR to follow convention. Such as `feat: add ability to create conditions`

And in your PR description you mentioned:

> As shown in numerous videos posted to, and discussed over, Teams.

Put link to those videos in PR

File: packages/ui/src/Utils/actionCondition.ts [hunk 3]
```diff
@@ -18,6 +17,13 @@ export const conditionDisplay: Record<CLM.ConditionType, string> = {
     [CLM.ConditionType.GREATER_THAN_OR_EQUAL]: '>=',
     [CLM.ConditionType.LESS_THAN]: '<',
     [CLM.ConditionType.LESS_THAN_OR_EQUAL]: '<=',
+    [CLM.ConditionType.STRING_EQUAL]: 'Matches exactly'
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/Utils/actionCondition.ts]: STRING_EQUAL is only available if and only if the selected comparison type is String (vs Numeric). This is enforced on the UI.  I suppose I could have reused ConditionType.Equal even for string comparisons, but then we'd have to have some logic on the Condition modal to display "Matches exactly" if the comparison type is string, and "==" if the comparison type is Numeric.  That would be way messier and more confusing IMO.   Adding this new ConditionType and having the UI select which ones are available based on ConditionType seems much cleaner and simpler IMO.

File: packages/ui/src/Utils/actionCondition.ts [hunk 4]
```diff
@@ -87,13 +97,31 @@ export const findNumberFromMemory = (memory: CLM.Memory, isMultivalue: boolean):
         return memory.entityValues.length
     }
 
-    const valueString: string | undefined = (memory?.entityValues?.[0]?.resolution as any).value
+    const valueString: string | undefined = (memory ?.entityValues ?.[0] ?.resolution as any).value
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/Utils/actionCondition.ts]: Must be some setting in VSCode.  Even if I get rid of the spaces, the minute I save, VSCode re-adds them.  I spent some time digging to find the setting and didn't find it.  In the interest of getting the PR sent out, I just left it.

File: packages/ui/src/Utils/actionCondition.ts [hunk 5]
```diff
@@ -87,13 +97,31 @@ export const findNumberFromMemory = (memory: CLM.Memory, isMultivalue: boolean):
...
         ? parseInt(valueString, 10)
         : undefined
 }
 
+/**
+ * Given memory value,
+ * return the userText field of the given memory value, if it exists, or undefined otherwise.
+ */
+export const findStringFromMemory = (memory: CLM.Memory): string | undefined => {
+    const valueString: string | null = (memory ?.entityValues ?.[0] ?.userText)
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/Utils/actionCondition.ts]: > Again remove spaces.

I'd love to, but VSCode won't let me... I'll keep digging and remove them if I find the setting...

[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/Utils/actionCondition.ts]: I had the method be of type string | undefined to have it be consistent with the other find*FromMemory methods.  The TS compiler complains (i.e., won't compile) if I have valueString be of type string | undefined.

File: packages/ui/src/components/modals/ConditionModal.tsx [hunk 6]
```diff
@@ -246,10 +341,17 @@ const Component: React.FC<Props> = (props) => {
         props.onClickCreate(theCondition)
     }
 
-    const isOperatorDisabled = selectedEntityOption?.data.entityType === CLM.EntityType.ENUM
-    const conditionsUsingEntity = props.conditions.filter(c => c.entityId === selectedEntityOption?.key)
+    var isOperatorDisabled = false
+    const setIsOperatorDisabled = () => {
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: because I use the same function later on.  I could have initialized it inline, but extracted into a function in the interest of encapsulation/reuse.

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 2 above)
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: >I would prefer that say "==" but not sure what you and @LarsLiden decided.
We had "decided" (to the extent that we agreed on Teams, and no one else objected) to have "Matches exactly" be the operator for string types.

>Also, when you select enum there seems to be bug with comparison type being undefined

For this one, it's not undefined, it's simply not a choice (currently) for enum types.  Again that's what was "decided"  over Teams (same as above... to the extent that it was discussed with @LarsLiden who agreed, and no one objected).

Comment (mattmazzola) [reply packages/ui/src/components/modals/ConditionModal.tsx]: Hmm, I thought you had proposed "matches exactly" before he proposed the design which introduced new "comparison type" dropdown and now that the dropdown was added, the new operator no longer needed to differential the types of equals.
Anyways, can let him confirm what he wanted.

> For this one, it's not undefined, it's simply not a choice (currently) for enum types. Again that's what was "decided" over Teams (same as above... to the extent that it was discussed with @LarsLiden who agreed, and no one objected).

This seems contradicting.  Not having a choice because it's constrained to single choice is not the same as there being no choices.  Example, when someone select string comparison type, you constraint the operators to "matches exactly" and disable the dropdown, yet here you use the argument that if the user can't pick any options, don't show any options.  It does not look good for dropdowns to have no value. It makes the UI look broken. (I think we also have this issue on Action Mocking modal since the options get removed as entities are added until they run out)

When enum entity is selected, the comparison type is similarly forced to be something. Currently it's string comparison type using value of ENUM, but could have a separate enum type if you want that.

Comment (mattmazzola) [reply packages/ui/src/components/modals/ConditionModal.tsx]: It occurred to me, that our required conditions are ANDed which means if you have contradition conditions the action will never be valid.
Is there any consideration where a user can choose options to disregard casing? 
"Matches Exactly" will mean "abc != Abc" and given the variations that people type in the bot, I would imagine it's a common to normalize to upper or lower.  Or is this not required for PVA integration?

Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: > It does not look good for dropdowns to have no value. It makes the UI look broken.... 
 When enum entity is selected, the comparison type is similarly forced to be something. Currently it's string comparison type using value of ENUM, but could have a separate enum type if you want that.

I can change it so that for enums, it's string comparison type.

File: packages/models/src/Action.ts [hunk 7]
```diff
@@ -24,12 +24,20 @@ export enum ConditionType {
...
+}
+
+export enum ComparisonType {
+  NUMBER_OF_ITEMS = "NUMBER_OF_ITEMS",
+  NUMERIC_VALUE = "NUMERIC_VALUE",
+  STRING = "STRING",
 }
 
 export interface Condition {
   entityId: string
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/models/src/Action.ts]: IIRC, the main reason(s) for using a stringValue field as opposed to a type field were:
- the usage of a type field was discussed in the past (before I even got here), and for whatever reason, was decided against. My assumption was that there was a concensus (or at least a compelling reason) not to use the type field, and was continuing in that pattern rather than revisiting past decisions.
- adding a type field would require a bunch of refactoring (not only here, but service side as well), that I wasn't really prepared to do at the time, since I was still getting up to speed on both this as well as service code.  Combining that refactoring effort along with these string matching changes wouldn't have been the best idea at the time.

If we do want to introduce a type field now that the condition logic is getting even richer than before, I certainly invite that discussion, but would say to do it separate from this PR.

File: packages/sdk/src/Utils.ts [hunk 8]
```diff
@@ -227,7 +227,10 @@ export const isConditionTrue = (
     filledEntities: CLM.FilledEntity[],
     entities: CLM.EntityBase[]
 ): boolean => {
-    if (condition.valueId) {
+    if (condition.stringValue){
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/sdk/src/Utils.ts]: done (in isStringConditionTrue).

File: packages/sdk/src/Utils.ts [hunk 9]
```diff
@@ -239,6 +242,21 @@ export const isConditionTrue = (
     }
 }
 
+const isStringConditionTrue = (
+    condition: CLM.Condition,
+    filledEntities: CLM.FilledEntity[]
+): boolean => {
+    if (condition.stringValue) {
+        const filledEntity = filledEntities.find(e => e.entityId === condition.entityId)
+        const stringValue = filledEntity && filledEntity.values.length > 0 && filledEntity.values[0].displayText
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/sdk/src/Utils.ts]: done.

File: packages/ui/src/Utils/actionCondition.ts [hunk 10]
```diff
@@ -18,6 +17,13 @@ export const conditionDisplay: Record<CLM.ConditionType, string> = {
     [CLM.ConditionType.GREATER_THAN_OR_EQUAL]: '>=',
     [CLM.ConditionType.LESS_THAN]: '<',
     [CLM.ConditionType.LESS_THAN_OR_EQUAL]: '<=',
+    [CLM.ConditionType.STRING_EQUAL]: 'Matches exactly'
+}
+
+export const comparisonTypeDisplay: Record<CLM.ComparisonType, string> = {
+    [CLM.ComparisonType.NUMBER_OF_ITEMS]: 'Number of Items',
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/Utils/actionCondition.ts]: Talked with @LarsLiden ; changing to values.

File: packages/ui/src/components/modals/ConditionModal.tsx [hunk 11]
```diff
@@ -73,6 +78,34 @@ const convertEnumValueToDropdownOption = (enumValue: CLM.EnumValue): EnumOption
     }
 }
 
+interface ComparisonTypeOption extends OF.IDropdownOption {
+    data: CLM.ComparisonType
+}
+
+const constructComparisonTypeDropdownOptions = (options: CLM.ComparisonType[]): ComparisonTypeOption[] => {
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: done

File: packages/ui/src/Utils/actionCondition.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (mattmazzola) [reply packages/ui/src/Utils/actionCondition.ts]: I don't understand your answer.  Also, please be more specific than "compiler won't have it". What error was it outputting? What value does compiler give to `valueString` if no explicit type is provided?

It is likely be case `userText` is defined as `userText: string | null`. 

Maybe `return memory?.entityValues?.[0]?.userText ?? undefined` will work.

File: packages/ui/src/Utils/actionCondition.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattmazzola) [reply packages/ui/src/Utils/actionCondition.ts]: Yes, later in the review I saw more spaces and realized it must be automated formatted but earlier in review I thought you were manually doing this. I think I saw this in some PVA code.  Maybe there is a user setting from those solutions that carried over into this one.

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 2 above)
Comment (mattmazzola) [reply packages/ui/src/components/modals/ConditionModal.tsx]: Yes, that's what I was asking for. It makes more sense to me to put something there. I think Enum or Strings. but since enums are basically using string equivalence. I thought we could re-use that.

File: packages/ui/src/components/modals/ConditionModal.tsx [hunk 12]
```diff
@@ -115,8 +171,42 @@ const Component: React.FC<Props> = (props) => {
...
+        if (!option) {
+            return
+        }
+
+        setSelectedComparisonType(option)
+        setOperatorOptions(option)
+    }
+
+    // Value text field
+    const onChangeTextField = (_: any, value: string) => {
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: done

File: packages/ui/src/components/modals/ConditionModal.tsx [hunk 13]
```diff
@@ -246,10 +341,17 @@ const Component: React.FC<Props> = (props) => {
         props.onClickCreate(theCondition)
     }
 
-    const isOperatorDisabled = selectedEntityOption?.data.entityType === CLM.EntityType.ENUM
-    const conditionsUsingEntity = props.conditions.filter(c => c.entityId === selectedEntityOption?.key)
+    var isOperatorDisabled = false
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: done

File: packages/ui/src/components/modals/ConditionModal.tsx [hunk 14]
```diff
@@ -246,10 +341,17 @@ const Component: React.FC<Props> = (props) => {
...
+    const setIsOperatorDisabled = () => {
+        isOperatorDisabled = (selectedEntityOption ?.data.entityType === CLM.EntityType.ENUM) || (selectedComparisonType === stringComparisonTypeOption)
+    }
+    setIsOperatorDisabled()
+    const conditionsUsingEntity = props.conditions.filter(c => c.entityId === selectedEntityOption ?.key)
     const currentCondition = createConditionFromState()
 
+
+    var isComparisonTypeDisabled = (selectedEntityOption ?.data.entityType === CLM.EntityType.ENUM)
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: done

File: packages/ui/src/components/modals/ConditionModal.tsx [hunk 15]
```diff
@@ -149,19 +240,21 @@ const Component: React.FC<Props> = (props) => {
...
-            setShowNumberValue(false)
-        }
-        else {
-            setShowNumberValue(true)
+            setShowStringField(false)
+        } else if (!entity.isMultivalue) {
+            setSelectedOperatorOption(stringEqualOperatorOption)
+            setShowStringField(true)
+            console.log(selectedComparisonType)
+        } else { // multivalue
```
Comment (mattmazzola) [inline packages/ui/src/components/modals/ConditionModal.tsx]: Prefer not to have inline comments. Would move to line above. Makes text manipulation and movement easier

Like:
```
...
}
// comment
else {
  ...
}
```

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (mattmazzola) [reply packages/ui/src/components/modals/ConditionModal.tsx]: Can you link to the line of "later on" case? I only see one call?

File: packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx [hunk 16]
```diff
@@ -174,15 +174,13 @@ class Container extends React.Component<Props, ComponentState> {
...
-
                     this.setState({
                         entityNameVal: nextProps.entity.entityName,
                         entityTypeVal: entityType,
                         entityResolverVal: resolverType,
                         isPrebuilt: isPrebuilt,
                         isMultivalueVal: nextProps.entity.isMultivalue,
-                        isMultiValueDisabled,
+                        isMultiValueDisabled: false,
```
Comment (mattmazzola) [inline packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx]: Based on comment above it seems it's not confirms how to deal with this, but if this property is always `false` then there is no need for the property. Can remove.

File: packages/ui/src/Utils/actionCondition.ts [hunk 17]
```diff
@@ -57,7 +67,7 @@ export const convertConditionToConditionalTag = (condition: CLM.Condition, entit
         }
     }
     else {
-        const name = getValueConditionName(entity, condition)
+        const name = condition.value ? getValueConditionName(entity, condition) : getStringConditionName(entity, condition)
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/Utils/actionCondition.ts]: done

File: packages/ui/src/Utils/actionCondition.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/Utils/actionCondition.ts]: found it & fixed...

File: packages/ui/src/Utils/actionCondition.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/Utils/actionCondition.ts]: and changed to userText

[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/Utils/actionCondition.ts]: but cleaned it up so that it's more succinct.

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: changed.

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: fixed

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 15 above)
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: sure, ok done.

File: packages/ui/src/components/modals/ConditionModal.tsx [hunk 18]
```diff
@@ -272,11 +374,19 @@ const Component: React.FC<Props> = (props) => {
                             <OF.Dropdown
                                 label="Entity"
                                 data-testid="condition-creator-modal-dropdown-entity"
-                                selectedKey={selectedEntityOption?.key}
+                                selectedKey={selectedEntityOption ?.key}
```
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: fixed

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: It was being called later on, but I guess it got refactored away at some point. I'll replace w/ the inline code instead.

File: packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx (hunk 16 above)
Comment (phui01) [reply packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx]: I put it back.

File: packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx (hunk 1 above)
Comment (phui01) [reply packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx]: I put it back.

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 2 above)
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: done

File: packages/sdk/src/Utils.ts [hunk 19]
```diff
@@ -110,7 +110,7 @@ const convertToMapById = (entityMap: CLM.FilledEntityMap): CLM.FilledEntityMap =
 export const actionHasHash = (actionId: string, hash: string, actions: CLM.ActionBase[]) => {
     // Check that correct action was taken
     const action = actions.find(a => a.actionId === actionId)
-    if (action?.clientData?.actionHashes) {
+    if (action ?.clientData ?.actionHashes) {
```
Comment (mattmazzola) [inline packages/sdk/src/Utils.ts]: Hmm, I see some of the spaces removed on other files but still see spaces here. I"m guessing it only fixed the files you opened and saved since you fixed the formatter. Also should update others to remove spaces.

File: packages/ui/src/components/modals/ConditionModal.tsx [hunk 20]
```diff
@@ -82,10 +82,10 @@ interface ComparisonTypeOption extends OF.IDropdownOption {
     data: CLM.ComparisonType
 }
 
-const constructComparisonTypeDropdownOptions = (options: CLM.ComparisonType[]): ComparisonTypeOption[] => {
+const convertComparisonTypeToDropdownOption = (options: CLM.ComparisonType[]): ComparisonTypeOption[] => {
     return options.map((comparisonType: string) => {
         let conditionText = `unknown`
-        if (comparisonTypeDisplay ?.[comparisonType]) {
+        if (comparisonTypeDisplay?.[comparisonType]) {
```
Comment (mattmazzola) [inline packages/ui/src/components/modals/ConditionModal.tsx]: Don't think you need the optional chaining here. `comparisonTypeDisplay` is a constant that will always be defined. Means remove the `?`

File: packages/ui/src/components/modals/ConditionModal.tsx [hunk 21]
```diff
@@ -98,13 +98,13 @@ const constructComparisonTypeDropdownOptions = (options: CLM.ComparisonType[]):
 }
 
 
-const singleEntityDropdownOptions = constructComparisonTypeDropdownOptions(Object.values(CLM.ComparisonType).filter(t => t !== CLM.ComparisonType.NUMBER_OF_ITEMS)) // number of items only valid for multi entities
-const multiEntityDropdownOptions = constructComparisonTypeDropdownOptions(Object.values(CLM.ComparisonType).filter(t => t === CLM.ComparisonType.NUMBER_OF_ITEMS)) // number of items only valid for multi entities
-const enumEntityDropdownOptions = constructComparisonTypeDropdownOptions([])
-var comparisonTypeDropdownOptions = singleEntityDropdownOptions
+const singleEntityDropdownOptions = convertComparisonTypeToDropdownOption(Object.values(CLM.ComparisonType).filter(t => t !== CLM.ComparisonType.NUMBER_OF_VALUES)) // number of items only valid for multi entities
+const multiEntityDropdownOptions = convertComparisonTypeToDropdownOption(Object.values(CLM.ComparisonType).filter(t => t === CLM.ComparisonType.NUMBER_OF_VALUES)) // number of items only valid for multi entities
+const enumEntityDropdownOptions = convertComparisonTypeToDropdownOption(Object.values(CLM.ComparisonType).filter(t => t === CLM.ComparisonType.STRING))
```
Comment (mattmazzola) [inline packages/ui/src/components/modals/ConditionModal.tsx]: I'm not sure where my old comment went about why we have these different lists. It seems unnecessary.

What was thoughts on consolidation and available options per comparison type.

File: packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx [hunk 22]
```diff
@@ -328,6 +330,13 @@ class Container extends React.Component<Props, ComponentState> {
         return newOrEditedEntity
     }
 
+    isMultiValueDisabled(entity: CLM.EntityBase, actions: CLM.ActionBase[]) {
```
Comment (mattmazzola) [inline packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx]: This code looks similar to the original code. I expected it to not show as green / addition if you were leaving it as it.

What's the change?

File: packages/ui/src/components/modals/ConditionModal.tsx [hunk 23]
```diff
@@ -148,20 +242,24 @@ const Component: React.FC<Props> = (props) => {
...
             setEnumValueOptions(valueOptions)
             // Only allow equal operator when selecting enum
-            setSelectedOperatorOption(equalOperatorOption)
-            setShowNumberValue(false)
+            setSelectedOperatorOption(stringEqualOperatorOption)
+            setShowStringField(false)
+        } else if (!entity.isMultivalue) {
+            setSelectedOperatorOption(stringEqualOperatorOption)
+            setShowStringField(true)
+            console.log(selectedComparisonType)
```
Comment (mattmazzola) [inline packages/ui/src/components/modals/ConditionModal.tsx]: remove the log

File: packages/ui/src/components/modals/ConditionModal.tsx [hunk 24]
```diff
@@ -246,10 +346,14 @@ const Component: React.FC<Props> = (props) => {
...
     }
 
-    const isOperatorDisabled = selectedEntityOption?.data.entityType === CLM.EntityType.ENUM
+    let isOperatorDisabled = (selectedEntityOption?.data.entityType === CLM.EntityType.ENUM) || (selectedComparisonType === stringComparisonTypeOption)
+
     const conditionsUsingEntity = props.conditions.filter(c => c.entityId === selectedEntityOption?.key)
     const currentCondition = createConditionFromState()
 
+
+    let isComparisonTypeDisabled = (selectedEntityOption?.data.entityType === CLM.EntityType.ENUM)
```
Comment (mattmazzola) [inline packages/ui/src/components/modals/ConditionModal.tsx]: seems like it should be `const`.  We only use `let` for things that need to be reassigned later. Same for other `let`s

File: packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx [hunk 25]
```diff
@@ -182,7 +182,7 @@ class Container extends React.Component<Props, ComponentState> {
                         entityResolverVal: resolverType,
                         isPrebuilt: isPrebuilt,
                         isMultivalueVal: nextProps.entity.isMultivalue,
-                        isMultiValueDisabled,
+                        isMultiValueDisabled: isMultiValueDisabled,
```
Comment (mattmazzola) [inline packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx]: Prefer short-hand these, don't need to duplicate the name. `isMultiValueDisabled,`

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 2 above)
Comment (mattmazzola) [reply packages/ui/src/components/modals/ConditionModal.tsx]: Ah here is he comment I was referring to.  It seems like you only made the comparison type string when enum entity is selected but didn't address the other parts of comment about changing list of options

File: packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx (hunk 22 above)
Comment (phui01) [reply packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx]: I think it's showing as a diff (green/added) as compared to the last commit?  I didn't change anything; just put the old code back in.

Comment (phui01) [reply packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx]: I imagine if I had squashed this commit in with the previous one, the diff would go away.  But I want(ed) to leave them separate so people can see what changes were made to address the previous comments

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 2 above)
Comment (mattmazzola) [reply packages/ui/src/components/modals/ConditionModal.tsx]: Also, on code reviews it's generally the responsibility of person who created comment to resolve.  GitHub's reviewing mechanism isn't as sophisticated as CodeFlow or ADO, so we don't have rules for github but it's kind of etiquette.

If it's simple and you're sure, then you can resolve but otherwise I'd ask.

Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: makes sense, sorry...

File: packages/sdk/src/Utils.ts (hunk 19 above)
Comment (mattmazzola) [reply packages/sdk/src/Utils.ts]: I think github comments can be stale sometimes, did you fix this?

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 20 above)
Comment (mattmazzola) [reply packages/ui/src/components/modals/ConditionModal.tsx]: What was resolution on this?

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 21 above)
Comment (mattmazzola) [reply packages/ui/src/components/modals/ConditionModal.tsx]: I thought you said we didn't need these different set of different options anymore since it's always constrained and disabled per my other.

File: packages/ui/src/components/modals/ConditionModal.tsx [hunk 26]
```diff
@@ -115,8 +174,42 @@ const Component: React.FC<Props> = (props) => {
         setSelectedOperatorOption(option)
     }
 
-    // Value
-    const [showNumberValue, setShowNumberValue] = React.useState(true)
+    // Comparison Type Dropdown
+
+    const setOperatorOptions = (option?: ComparisonTypeOption) => {
+        if (option === stringComparisonTypeOption) {
+            operatorOptions = stringEqualOperatorOptions
```
Comment (mattmazzola) [inline packages/ui/src/components/modals/ConditionModal.tsx]: I think since you changed it to use `allOperatorOptions` instead of `operatorOptions` you can remove this

File: packages/sdk/src/Utils.ts (hunk 19 above)
Comment (mattmazzola) [reply packages/sdk/src/Utils.ts]: What's the status here, I thought you were wanting to get this merged?  I don't see this comment as outdated and I still see bad formatting.  If someone else opens this file later these will show as changes they made because of autoformat even though they didn't change them.

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 23 above)
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: removed

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 24 above)
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: changed

File: packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx (hunk 25 above)
Comment (phui01) [reply packages/ui/src/components/modals/EntityCreatorEditor/EntityCreatorContainer.tsx]: done

File: packages/sdk/src/Utils.ts (hunk 19 above)
Comment (phui01) [reply packages/sdk/src/Utils.ts]: fixed

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 20 above)
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: removed

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 21 above)
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: No, we still need these, since the set of available options differs depending on the entity type.

File: packages/ui/src/components/modals/ConditionModal.tsx (hunk 26 above)
Comment (phui01) [reply packages/ui/src/components/modals/ConditionModal.tsx]: No, we still need this; the set of available operator options depends on the comparison type.
